test/remote/instance.go

```diff
@@ -121,31 +123,47 @@ func (i *InstanceInfo) CreateOrGetInstance(imageURL, serviceAccount string) erro
 		Email:  serviceAccount,
 		Scopes: []string{"https://www.googleapis.com/auth/cloud-platform"},
 	}
-	inst.ServiceAccounts = []*compute.ServiceAccount{saObj}
+	newInst.ServiceAccounts = []*compute.ServiceAccount{saObj}
 
 	if pubkey, ok := os.LookupEnv("JENKINS_GCE_SSH_PUBLIC_KEY_FILE"); ok {
 		klog.V(4).Infof("JENKINS_GCE_SSH_PUBLIC_KEY_FILE set to %v, adding public key to Instance", pubkey)
 		meta, err := generateMetadataWithPublicKey(pubkey)
 		if err != nil {
 			return err
 		}
-		inst.Metadata = meta
+		newInst.Metadata = meta
 	}
 
-	if _, err := i.computeService.Instances.Get(i.project, i.zone, inst.Name).Do(); err != nil {
-		op, err := i.computeService.Instances.Insert(i.project, i.zone, inst).Do()
-		klog.V(4).Infof("Inserted instance %v in project: %v, zone: %v", inst.Name, i.project, i.zone)
-		if err != nil {
-			ret := fmt.Sprintf("could not create instance %s: API error: %v", i.name, err)
-			if op != nil {
-				ret = fmt.Sprintf("%s. op error: %v", ret, op.Error)
+	curInst, _ := i.computeService.Instances.Get(i.project, i.zone, newInst.Name).Do()
+	if curInst != nil {
+		if !strings.Contains(curInst.MachineType, newInst.MachineType) {
+			klog.V(4).Infof("Instance machine type doesn't match the required one. Remove instance.")
+			if _, err := i.computeService.Instances.Delete(i.project, i.zone, i.name).Do(); err != nil {
+				return err
+			}
+
+			then := time.Now()
+			err := wait.Poll(15*time.Second, 5*time.Minute, func() (bool, error) {
+				klog.V(2).Infof("Waiting for instance to be deleted. %v elapsed", time.Since(then))
+				if instance, _ = i.computeService.Instances.Get(i.project, i.zone, i.name).Do(); instance != nil {
+					return false, nil
+				}
+				return true, nil
+			})
+			if err != nil {
+				return err
 			}
-			return errors.New(ret)
-		} else if op.Error != nil {
-			return fmt.Errorf("could not create instance %s: %+v", i.name, op.Error)
+
+			if err := insertInstance(i, newInst); err != nil {
```

Comment:
Can you refactor this so your function only calls insertInsert 1x instead of 2x? This if statement is 3 levels deep at one point and I think the logic can be simplified to escape out of some error conditions prior to returning success. IMO the main flow of a function should be the happy-path, with if statements reserved for exceptional scenarios and error handling. I think the general flow can be:
 * Check if instance exists. If instance exists and not matching machine-type: Delete and verify instance does not exist.
 * If instance not exists: Create

Reply:
Make sense. I remove the insertInstance method and add the original code block back. Check the machine-type before insertion.